src/engine/MultiColumnJoin.cpp

```diff
@@ -0,0 +1,471 @@
+  for (size_t i = 0; i < _joinColumns.size(); i++) {
+    size_t dl = std::max(1.0f, _left->getSizeEstimate() /
+                                   _left->getMultiplicity(_joinColumns[i][0]));
+    size_t dr = std::max(1.0f, _right->getSizeEstimate() /
+                                   _right->getMultiplicity(_joinColumns[i][1]));
+    numDistinctLeft = std::min(numDistinctLeft, dl);
+    numDistinctRight = std::min(numDistinctRight, dr);
+  }
+  size_t numDistinctResult = std::min(numDistinctLeft, numDistinctRight);
+
+  // compute an estimate for the results multiplicity
+  float multLeft = std::numeric_limits<float>::max();
+  float multRight = std::numeric_limits<float>::max();
+  for (size_t i = 0; i < _joinColumns.size(); i++) {
+    multLeft = std::min(multLeft, _left->getMultiplicity(_joinColumns[i][0]));
+    multRight = std::min(multLeft, _left->getMultiplicity(_joinColumns[i][1]));
+  }
+  float multResult = multLeft * multRight;
+
+  _sizeEstimate = multResult * numDistinctResult;
+  // Don't estimate 0 since then some parent operations
+  // (in particular joins) using isKnownEmpty() will
+  // will assume the size to be exactly zero
+  _sizeEstimate += 1;
+
+  // compute estimates for the multiplicities of the result columns
+  _multiplicities.clear();
+
+  for (size_t i = 0; i < _left->getResultWidth(); i++) {
+    float mult = _left->getMultiplicity(i) * (multResult / multLeft);
+    _multiplicities.push_back(mult);
+  }
+
+  for (size_t i = 0; i < _right->getResultWidth(); i++) {
+    bool isJcl = false;
+    for (size_t j = 0; j < _joinColumns.size(); j++) {
+      if (_joinColumns[j][1] == i) {
+        isJcl = true;
+        break;
+      }
+    }
+    if (isJcl) {
+      continue;
+    }
+    float mult = _right->getMultiplicity(i) * (multResult / multRight);
+    _multiplicities.push_back(mult);
+  }
+  _multiplicitiesComputed = true;
+}
+
+template <typename R>
+struct MultiColumnJoinResultAllocator {
+  static R allocate(size_t width) {
+    (void)width;
+    return R();
+  }
+};
+
+template <>
+struct MultiColumnJoinResultAllocator<std::vector<Id>> {
+  static std::vector<Id> allocate(size_t width) {
+    return std::vector<Id>(width);
+  }
+};
+
+template <typename A, typename B, typename R>
+void MultiColumnJoin::computeMultiColumnJoin(
+    const A& a, const B& b, const vector<array<Id, 2>>& joinColumns,
+    vector<R>* result, size_t resultWidth) {
+  // check for trivial cases
+  if (a.size() == 0 || b.size() == 0) {
+    return;
+  }
+
+  // Marks the columns in b that are join columns. Used to skip these
+  // when computing the result of the join
+  int joinColumnBitmap_b = 0;
+  for (const array<Id, 2>& jc : joinColumns) {
+    joinColumnBitmap_b |= (1 << jc[1]);
+  }
+
+  // Cast away constness so we can add sentinels that will be removed
+  // in the end and create and add those sentinels.
+  A& l1 = const_cast<A&>(a);
+  B& l2 = const_cast<B&>(b);
+  Id sentVal = std::numeric_limits<Id>::max() - 1;
```

Comment:
I know it's in the original code but I'm still a little skeptical if the use of sentinels actually improves the performance. Especially since removing the constness may hinder compiler opts and since th `push_back()` may trigger allocation unless it's always considered in `reserve()`. Maybe add a TODO to clear this up with performance tests.

Reply:
I added the TODO.